src/tools/fuzzing/heap-types.cpp

```diff
@@ -0,0 +1,499 @@
+      switch (type) {
+        case HeapType::ext:
+        case HeapType::i31:
+          // No other subtypes.
+          return type;
+        case HeapType::func:
+          return generateSignature();
+        case HeapType::any:
+          return generateSubAny();
+        case HeapType::eq:
+          return generateSubEq();
+        case HeapType::data:
+          return generateSubData();
+      }
+      WASM_UNREACHABLE("unexpected index");
+    }
+  }
+
+  template<typename Kind> std::optional<HeapType> pickKind() {
+    std::vector<HeapType> candidates;
+    // Iterate through the top level kinds, finding matches for `Kind`.
+    for (Index i = 0; i < typeKinds.size(); ++i) {
+      if (std::get_if<Kind>(&typeKinds[i])) {
+        candidates.push_back(builder[i]);
+      }
+    }
+    if (candidates.size()) {
+      return rand.pick(candidates);
+    } else {
+      return {};
+    }
+  }
+
+  HeapType pickSubFunc() {
+    if (auto type = pickKind<SignatureKind>()) {
+      return *type;
+    } else {
+      return HeapType::func;
+    }
+  }
+
+  HeapType pickSubData() {
+    if (auto type = pickKind<DataKind>()) {
+      return *type;
+    } else {
+      return HeapType::data;
+    }
+  }
+
+  HeapType pickSubEq() {
+    if (rand.oneIn(2)) {
+      return HeapType::i31;
+    } else {
+      return pickSubData();
+    }
+  }
+
+  HeapType pickSubAny() {
+    switch (rand.upTo(4)) {
+      case 0:
+        return HeapType::func;
+      case 1:
+        return HeapType::eq;
+      case 2:
+        return pickSubFunc();
+      case 3:
+        return pickSubEq();
+    }
+    WASM_UNREACHABLE("unexpected index");
+  }
+
+  HeapType pickSubHeapType(HeapType type) {
+    auto it = typeIndices.find(type);
+    if (it != typeIndices.end()) {
+      // This is a constructed type, so we know where its subtypes are.
+      return builder[rand.pick(subtypeIndices[typeIndices[type]])];
+    } else {
+      // This is not a constructed type, so it must be a basic type.
+      assert(type.isBasic());
+      switch (type.getBasic()) {
+        case HeapType::func:
+          return pickSubFunc();
+        case HeapType::ext:
+          return HeapType::ext;
+        case HeapType::any:
+          return pickSubAny();
+        case HeapType::eq:
+          return pickSubEq();
+        case HeapType::i31:
+          return HeapType::i31;
+        case HeapType::data:
+          return pickSubData();
+      }
+      WASM_UNREACHABLE("unexpected kind");
+    }
+  }
+
+  // TODO: Make this part of the wasm-type.h API
+  struct Ref {
+    HeapType type;
+    Nullability nullability;
+  };
+
+  Ref generateSubRef(Ref super) {
+    auto nullability = super.nullability == NonNullable
+                         ? NonNullable
+                         : rand.oneIn(2) ? Nullable : NonNullable;
+    return {pickSubHeapType(super.type), nullability};
+  }
+
+  Rtt generateSubRtt(Rtt super) {
+    auto depth = super.hasDepth()
+                   ? super.depth
```

Comment:
I don't remember there ever being a rule for subtyping between rtts of different depths. This makes sense because the rtt value in the engine must have some particular depth, so it would be unsound to type that value with two different depths.